### PR TITLE
fix(git,forge): ensure deterministic credential helper for HTTPS remotes on macOS

### PR DESCRIPTION
**Pull request by theBGuy**
This change ensures that when the app is launched via Finder or the Dock on macOS, network operations against private HTTPS repositories (such as fetch, pull, and push) consistently use an absolute path to the required credential helper (`gh`/`glab`). This fixes failures where git is unable to authenticate due to a minimal PATH, resulting in "Repository not found" errors for private repos. Git commands now inject the resolved absolute helper path, so authentication works regardless of launch context.

### Git credential handling
- Adds `with_credentials` helper in `src-tauri/src/git/remote.rs` to prepend one-shot credential helper args to git commands.
- Updates `git_fetch_core`, `git_fetch_remote`, `git_remote_default_branch`, `git_pull_core`, and `git_push_core` in `src-tauri/src/git/remote.rs` to use provider-specific credential helpers determined at runtime.
- Integrates `credential_config_for_remote` to derive appropriate helper configs for each git remote in `src-tauri/src/forge/mod.rs`.

### Forge credential resolution
- Implements `clone_credential_config` and `github_credential_entry` in `src-tauri/src/forge/github.rs` to resolve and inject the absolute path to `gh` for HTTPS remote URLs.
- Mirrors this logic with centralized credential resolution in `credential_config_for_remote` in `src-tauri/src/forge/mod.rs`, handling GitHub, GitLab, and Bitbucket accordingly.
- Ensures that resolution fails open (does not break ambient helpers) if required CLIs are absent.

### Utility and tests
- Adds helper function `is_https_remote` in `src-tauri/src/forge/mod.rs` to distinguish HTTPS from SSH URLs.
- Expands tests in both `github.rs` and `mod.rs` to cover new helpers and edge cases.

### Documentation
- Adds `changelog.d/fixed-macos-fetch-repo-not-found.md` documenting the fixed issue with macOS app launches breaking private repo network operations.

### CODE REVIEW

**Comment by theBGuy** [comment]
🤖 **GitDesktop AI review** · `sonnet` · automated

---

This PR correctly addresses the macOS GUI-launch PATH issue by injecting absolute credential helper paths as one-shot `-c` args, and the core `credential_config_for_remote` fail-open design is sound. However, there's one concrete regression and one architectural inconsistency worth addressing before merging.

### Correctness

**blocker** — `forge_clone` / `src-tauri/src/forge/mod.rs:602` — GitHub clone hard-fails when `gh` isn't installed

Before this PR, `forge_clone` for GitHub fell into `_ => Vec::new()`, passing no extra args and relying on ambient auth (git-credential-manager, keychain, etc.). After this PR:

```rust
Provider::GitHub => github::clone_credential_config(&url).await?,
```

`github::clone_credential_config` does `.ok_or(AppError::GhNotFound)?` — so if `gh` isn't found, the `?` propagates `GhNotFound` all the way to the caller. Any user who authenticates via git-credential-manager or another ambient mechanism rather than `gh` can no longer clone GitHub repos through the UI. The PR description says "resolution fails open if required CLIs are absent" — but that's only true in `credential_config_for_remote` (which uses `.unwrap_or_default()`); `forge_clone` doesn't.

Concrete trigger: user with git-credential-manager installed but without `gh` clicks Clone on a private GitHub repo → "GitHub CLI (gh) not found."

Fix: apply the same fail-open pattern in `forge_clone`:
```rust
Provider::GitHub => github::clone_credential_config(&url).await.unwrap_or_default(),
```

Or, cleaner, change `clone_credential_config` to return `Ok(Vec::new())` when the CLI is absent (and let a distinct error variant be the caller's choice), matching how `credential_config_for_remote` already treats absence.

---

### Logic

**should-fix** — `credential_config_for_remote` / `src-tauri/src/forge/mod.rs:157–161` — provider detected from `origin` but credentials applied to the requested remote's URL

```rust
let url = git_remote_url(repo_path, remote).await...;  // fetched from 'remote'
...
match detect_non_github(repo_path).await {              // always reads 'origin' internally
    Some((Provider::GitLab, _)) => gitlab::clone_credential_config(&url)...
    ...
}
```

`detect_non_github` unconditionally reads `origin`'s URL. If `remote != "origin"` (the `git_fetch_remote` / `git_remote_default_branch` path uses caller-supplied remotes), and origin and that remote are on different forges — e.g., `origin` = `gitlab.com`, `upstream` = `github.com` — then the URL handed to `gitlab::clone_credential_config` is the GitHub URL, producing `credential.https://github.com.helper=!"/glab" auth git-credential`. Git probably falls through since glab ignores github.com, but this is wrong by construction.

Fix: drive provider detection off `url`'s own host rather than `detect_non_github`:
```rust
let host = remote_host(&url).unwrap_or_default();
let provider = provider_for_host(&host)
    .or_else(|| glab::known_hosts_sync().contains(&host).then_some(Provider::GitLab));
match provider { ... }
```

---

### Readability

**nit** — `src-tauri/src/git/remote.rs`, all five call sites — The `.iter().map(String::as_str).collect::<Vec<_>>()` dance is repeated verbatim at every call site because `with_credentials` returns `Vec<String>` but `run_git_mutating` takes `&[&str]`. Consider giving `with_credentials` a companion that also does the conversion, or accepting `&[impl AsRef<str>]` throughout. Not a bug, but five identical boilerplate chains are an invitation for a future divergence.

---

_Posted by [GitDesktop](https://gitdesktop.app) — AI output, verify before acting on it._

**Comment by theBGuy** [comment]
🤖 **GitDesktop AI security audit** · `sonnet` · automated

---

No security issues meeting the reporting threshold.

The credential injection mechanism is sound: `with_credentials` appends `-c key=value` entries as proper argv elements (not shell-split strings), so there is no process-level shell injection. The `host` value extracted by `remote_host` (split on `/` and `:` only) lands in the git config *key* (`credential.https://{host}.helper`), not the *value* that git shell-executes via `!`. Even with an adversarially crafted remote URL containing `=` in the hostname, the first `=` in the combined string would simply shift the key/value split point — the resulting value would not begin with `!`, so git would attempt to exec a nonexistent binary rather than invoke a shell. The `gh_path` / `glab_path` values come from `resolve_named`, a trusted OS-level binary lookup, and on supported platforms cannot contain `"` or other shell metacharacters that would break out of the surrounding double-quote wrapper.

The `is_https_remote` accepting `http://` is a logical inconsistency (the generated `credential.https://…` entry does not match `http://` remotes in git's URL matching), but it is a no-op with no credential leakage.

The `forge_clone` regression flagged in the prior review comment — `github::clone_credential_config` propagating `GhNotFound` via `?` instead of failing open with `.unwrap_or_default()` — remains unfixed in this diff, but it is an availability regression, not an exploitable security vulnerability (it breaks clone for users without `gh`, it does not expose credentials or bypass auth).

---

_Posted by [GitDesktop](https://gitdesktop.app) — AI output, verify before acting on it._

**Comment by theBGuy** [comment]
**Round-1 dispositions** (all four items accepted; fixes land in the next push):

1. **Blocker — `forge_clone` GitHub arm fails closed: accepted, fixed.** The arm now degrades `.unwrap_or_default()` when gh can't be resolved, restoring pre-PR ambient-auth clone behavior (public repos need no auth; a GCM user already has HTTPS auth). GitLab's arm deliberately stays strict — glab *is* GitLab's auth path here, and that strictness predates this PR. Good catch: the "clone stays strict" call treated the two providers as symmetric when only GitLab's strictness was pre-existing behavior.

2. **Should-fix — provider detected from `origin` while credentials apply to the requested remote: accepted, fix re-derived.** The finding is right (a cross-forge origin/upstream pair got the wrong CLI's helper on the `git_fetch_remote`/`set-head` paths). One correction to the suggested fix: `glab::known_hosts_sync()` doesn't exist — the real API is the async `glab::known_hosts()` (forge/glab.rs), which is what `detect_non_github` itself uses. Implemented as a pure, unit-tested `provider_for_remote_host(host, glab_hosts)` classifying the *requested* remote's own host (canonical hosts direct; glab-known custom hosts → self-managed GitLab; github.com/GHE/unknown → gh-default routing), with the glab config read skipped for canonical hosts. Identical outcome for `origin`; correct for cross-forge remotes; one fewer origin resolution. New test covers gitlab.com / bitbucket.org / github.com / glab-known custom host / unknown.

3. **Nit — five repeated `as_str` conversion chains: accepted.** Collapsed into one module-private `run_git_mutating_with_creds(state, repo, cred, sub, timeout)`; all five network sites now route through it.

4. **Security audit's `http://` note: accepted as a trivial tightening.** `is_https_remote` is now `https://`-only — the helper entry we format keys on `credential.https://…` and would never match an http remote, so the old acceptance was a wasted resolve + a misleading gate. Doc + test updated.

Verification on the fixed tree: `cargo test` 656 passed / 0 failed (incl. the new cross-forge classification test), `cargo clippy -- -D warnings` clean. The macOS end-to-end verify-gap stands as described in the PR body — the 3-OS CI legs compile-check the fix on mac, but the Finder-launch auth behavior still needs a live mac run.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

**Comment by theBGuy** [comment]
🤖 **GitDesktop AI review** · `sonnet` · automated

---

Both previous findings are addressed in this round and no new issues are visible in the current diff.

### Resolved since last review

- **Blocker — `forge_clone` GitHub arm hard-fails when `gh` absent.** Fixed: `Provider::GitHub => github::clone_credential_config(&url).await.unwrap_or_default()`. The GitLab arm correctly stays strict (`?`) since glab is GitLab's only auth path; GitHub degrades gracefully to ambient auth. ✓

- **Should-fix — provider detected from `origin` while credentials applied to the requested remote's URL.** Fixed: `credential_config_for_remote` now extracts the host from the *requested* remote's own URL, classifies it via the new pure `provider_for_remote_host(host, glab_hosts)` helper (mirroring `detect_non_github`'s logic without the hard-wired `origin` read), and skips the `glab::known_hosts()` I/O entirely for canonical hosts. New unit tests cover all four cases. ✓

- **Nit — repeated `.iter().map(String::as_str).collect::<Vec<_>>()` at five call sites.** Eliminated by `run_git_mutating_with_creds`, which centralises both the `with_credentials` composition and the `String → &str` slice conversion. ✓

**Copilot's six "borrow of temporary across `.await`" findings** were all against the old call sites; those call sites no longer exist. The remaining instance — inside `run_git_mutating_with_creds` itself — is not a compile error: the temporary `Vec<&str>` is created in function-argument position, so its lifetime extends to the end of the containing `.await` expression per Rust's temporary-lifetime extension rules. Not an issue.

The implementation is correct, the fail-open / fail-strict split is now principled and consistent across both `credential_config_for_remote` and `forge_clone`, and the new helper is well-tested. Ready to merge.

---

_Posted by [GitDesktop](https://gitdesktop.app) — AI output, verify before acting on it._

File: src-tauri/src/git/remote.rs [hunk 1]
```diff
@@ -132,7 +143,15 @@ pub async fn git_fetch(state: State<'_, AppState>, repo_path: String) -> AppResu
...
-    run_git_mutating(state, &repo_path, &["fetch", "--prune"], NETWORK_TIMEOUT).await?;
+    let cred = crate::forge::credential_config_for_remote(&repo_path, "origin").await?;
+    let args = with_credentials(&cred, &["fetch", "--prune"]);
+    run_git_mutating(
+        state,
+        &repo_path,
+        &args.iter().map(String::as_str).collect::<Vec<_>>(),
+        NETWORK_TIMEOUT,
+    )
+    .await?;
```
[the remark this answers is not in the document]
Comment (theBGuy) [reply src-tauri/src/git/remote.rs]: Refuted, then made moot. A temporary created in function-argument position lives until the end of the enclosing statement — including through the `.await` on that same expression — so this was never a compile error, and empirically the flagged code built green (`cargo build` + full test suite + `clippy -D warnings`) before any change. As of e6aa77a the five call sites are consolidated into `run_git_mutating_with_creds` anyway, so the pattern now exists in exactly one audited place.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

File: src-tauri/src/git/remote.rs [hunk 2]
```diff
@@ -165,10 +184,12 @@ pub async fn git_fetch_remote(
...
 ) -> AppResult<()> {
     ensure_remote_exists(&repo_path, &remote).await?;
+    let cred = crate::forge::credential_config_for_remote(&repo_path, &remote).await?;
+    let args = with_credentials(&cred, &["fetch", "--prune", &remote]);
     run_git_mutating(
         &state,
         &repo_path,
-        &["fetch", "--prune", &remote],
+        &args.iter().map(String::as_str).collect::<Vec<_>>(),
         NETWORK_TIMEOUT,
```
[the remark this answers is not in the document]
Comment (theBGuy) [reply src-tauri/src/git/remote.rs]: Same disposition as the sibling threads: refuted (argument-position temporaries live through the enclosing `.await` statement; the code compiled and passed the full suite as written) and moot since e6aa77a consolidated the call sites into `run_git_mutating_with_creds`.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

File: src-tauri/src/git/remote.rs [hunk 3]
```diff
@@ -204,10 +225,12 @@ pub async fn git_remote_default_branch(
...
     // The local ref is unset — ask the remote for its HEAD (one network call),
     // then re-read. `set-head --auto` writes `refs/remotes/<remote>/HEAD`.
+    let cred = crate::forge::credential_config_for_remote(&repo_path, &remote).await?;
+    let args = with_credentials(&cred, &["remote", "set-head", &remote, "--auto"]);
     run_git_mutating(
         &state,
         &repo_path,
-        &["remote", "set-head", &remote, "--auto"],
+        &args.iter().map(String::as_str).collect::<Vec<_>>(),
         NETWORK_TIMEOUT,
```
[the remark this answers is not in the document]
Comment (theBGuy) [reply src-tauri/src/git/remote.rs]: Same disposition as the sibling threads: refuted (argument-position temporaries live through the enclosing `.await` statement; the code compiled and passed the full suite as written) and moot since e6aa77a consolidated the call sites into `run_git_mutating_with_creds`.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

File: src-tauri/src/git/remote.rs [hunk 4]
```diff
@@ -268,7 +291,15 @@ pub(crate) async fn git_pull_core(
...
-    run_git_mutating(state, &repo_path, &["pull", flag], NETWORK_TIMEOUT).await?;
+    let cred = crate::forge::credential_config_for_remote(&repo_path, "origin").await?;
+    let args = with_credentials(&cred, &["pull", flag]);
+    run_git_mutating(
+        state,
+        &repo_path,
+        &args.iter().map(String::as_str).collect::<Vec<_>>(),
+        NETWORK_TIMEOUT,
+    )
+    .await?;
```
[the remark this answers is not in the document]
Comment (theBGuy) [reply src-tauri/src/git/remote.rs]: Same disposition as the sibling threads: refuted (argument-position temporaries live through the enclosing `.await` statement; the code compiled and passed the full suite as written) and moot since e6aa77a consolidated the call sites into `run_git_mutating_with_creds`.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

File: src-tauri/src/git/remote.rs [hunk 5]
```diff
@@ -296,7 +327,15 @@ pub(crate) async fn git_push_core(
...
-    run_git_mutating(state, &repo_path, &args, NETWORK_TIMEOUT).await?;
+    let cred = crate::forge::credential_config_for_remote(&repo_path, "origin").await?;
+    let args = with_credentials(&cred, &args);
+    run_git_mutating(
+        state,
+        &repo_path,
+        &args.iter().map(String::as_str).collect::<Vec<_>>(),
+        NETWORK_TIMEOUT,
+    )
+    .await?;
```
[the remark this answers is not in the document]
Comment (theBGuy) [reply src-tauri/src/git/remote.rs]: Same disposition as the sibling threads: refuted (argument-position temporaries live through the enclosing `.await` statement; the code compiled and passed the full suite as written) and moot since e6aa77a consolidated the call sites into `run_git_mutating_with_creds`.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

File: src-tauri/src/forge/mod.rs [hunk 6]
```diff
@@ -133,6 +133,41 @@ async fn detect_non_github(repo_path: &str) -> Option<(Provider, String)> {
...
+        return Ok(Vec::new()); // SSH → keys, not helpers
+    }
+    // Fail open when the CLI can't be resolved: a user with working ambient auth
+    // (e.g. git-credential-manager) must not have every HTTPS fetch/pull/push hard-
+    // fail just because gh/glab isn't installed.
+    match detect_non_github(repo_path).await {
+        Some((Provider::GitLab, _)) => Ok(gitlab::clone_credential_config(&url).await.unwrap_or_default()),
+        Some((Provider::Bitbucket, _)) => Ok(Vec::new()),
+        _ => Ok(github::clone_credential_config(&url).await.unwrap_or_default()),
+    }
```
[the remark this answers is not in the document]
Comment (theBGuy) [reply src-tauri/src/forge/mod.rs]: Accepted — this converged with the AI review's should-fix and is fixed in e6aa77a: provider detection now classifies the *requested* remote's own host via the pure, unit-tested `provider_for_remote_host(host, glab_hosts)` (canonical hosts direct; glab-known custom hosts → self-managed GitLab; github.com/GHE/unknown → gh-default routing), which also drops the redundant `git remote get-url origin` this thread noted. Cross-forge origin/upstream pairs each get their own CLI's helper.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._
